File: src/main/java/no/bekk/java/exercises/Ex7_ExtendedTypes.java
package no.bekk.java.exercises;

import no.bekk.java.model.Car;

import java.util.List;

interface Competing<T> {
	T get();
	Competing<T> competeWith(Competing<T> other);
}
public class Ex7_ExtendedTypes {
	static public <T> T compete(List<Competing<T>> competitors) {
		return competitors.stream().reduce((x, y) -> x.competeWith(y)).get().get(); // TODO: Remove for excercise
	}
}

// TODO: Remove for excercise
class CarAdapter implements Competing<Car> {
	private Car car;

	CarAdapter(Car car) {
		this.car = car;
	}

	public Car get() {return car;}

	public Competing<Car> competeWith(Competing<Car> other) {
		return car.hp > other.get().hp ? this : other;
	}
}

File: src/main/scala/no/bekk/scala/exercises/Ex7_ExtendedTypes.scala
package no.bekk.scala.exercises

import no.bekk.scala.exercises.Ex7_ExtendedTypes.Competing
import no.bekk.scala.model.Car

object Ex7_ExtendedTypes {

	trait Competing[T] {
		def compete(a: T, b: T): T
	}

	def compete[A](competitors: List[A])(implicit ev: Competing[A]): A = ???
}

object CompetingTypes {
	implicit object CompetingCar extends Competing[Car] {
		def compete(a: Car, b: Car): Car = ???
	}
}